net/libfab/libfab.c

```diff
@@ -0,0 +1,2963 @@
+	   Hence, all such events can be ignored.
+	*/
+	return 0;
+}
+
+/**
+ * Check connection established and shutdown events for transmit endpoint.
+ * All other types of events are ignored.
+ */
+static void libfab_txep_event_check(struct m0_fab__ep *txep,
+				    struct m0_fab__active_ep *aep,
+				    struct m0_fab__tm *tm)
+{
+	uint32_t rc;
+
+	rc = libfab_check_for_event(aep->aep_tx_res.ftr_eq);
+	if (rc == FI_CONNECTED) {
+		libfab_pending_bufs_send(txep);
+		aep->aep_tx_state = FAB_CONNECTED;
+	} else if (rc == FI_SHUTDOWN) {
+		/* Reset and reopen endpoint */
+		libfab_txep_init(aep, tm, txep);
+	}
+	/* All other types of events can be ignored */
+}
+
+/**
+ * Check for completion events on the completion queue for the receive endpoint
+ */
+static void libfab_rxep_comp_read(struct fid_cq *cq, struct m0_fab__ep *ep)
+{
+	struct m0_fab__buf *buf[FAB_MAX_COMP_READ];
+	m0_bindex_t         len[FAB_MAX_COMP_READ];
+	uint64_t            data[FAB_MAX_COMP_READ];
+	int                 i;
+	int                 cnt;
+
+	memset(&buf, 0, sizeof(buf));
+	memset(&len, 0, sizeof(len));
+	if (cq != NULL) {
+		cnt = libfab_check_for_comp(cq, buf, len, data);
+		for (i = 0; i < cnt; i++) {
+			if (buf[i] != NULL) {
+				if (buf[i]->fb_length == 0)
+					buf[i]->fb_length = len[i];
+				buf[i]->fb_ev_ep = ep;
+				libfab_buf_done(buf[i], 0);
+			}
+			if (data[i] != 0)
+				libfab_buf_done((struct m0_fab__buf*)data[i],
+						0);
+		}
+	}
+}
+
+/**
+ * Check for completion events on the completion queue for the transmit endpoint
+ */
+static void libfab_txep_comp_read(struct fid_cq *cq)
+{
+	struct m0_fab__active_ep *aep;
+	struct m0_fab__buf       *buf[FAB_MAX_COMP_READ];
+	int                       i;
+	int                       cnt;
+
+	memset(&buf, 0, sizeof(buf));
+	cnt = libfab_check_for_comp(cq, buf, NULL, NULL);
+	for (i = 0; i < cnt; i++) {
+		if (buf[i] != NULL) {
+			aep = libfab_aep_get(buf[i]->fb_txctx);
+			if (M0_IN(buf[i]->fb_nb->nb_qtype,
+				  (M0_NET_QT_ACTIVE_BULK_RECV,
+				   M0_NET_QT_ACTIVE_BULK_SEND)))
+				--aep->aep_bulk_cnt;
+			buf[i]->fb_wr_comp_cnt++;
+			if (buf[i]->fb_wr_comp_cnt >= buf[i]->fb_wr_cnt) {
+				libfab_target_notify(buf[i], aep);
+				libfab_buf_done(buf[i], 0);
+			}
+		}
+	}
+}
+
+/**
+ * Used to poll for connection events, completion events and process the queued
+ * bulk buffer operations.
+ */
+static void libfab_poller(struct m0_fab__tm *tm)
+{
+	struct m0_fab__ev_ctx    *ctx;
+	struct m0_fab__ep        *xep;
+	struct m0_fab__active_ep *aep;
+	struct fid_cq            *cq;
+	struct epoll_event        ev;
+	int                       ev_cnt;
+
+	/* Wait for the transfer machine to get started */
+	while (tm->ftm_state != FAB_TM_STARTED)
+		/* No-op */;
+	
+	while (tm->ftm_state != FAB_TM_SHUTDOWN) {
+		sched_yield();
+		ev_cnt = epoll_wait(tm->ftm_epfd, &ev, 1, FAB_WAIT_FD_TMOUT);
+
+		while (1) {
+			m0_mutex_lock(&tm->ftm_endlock);
+			if (tm->ftm_state == FAB_TM_SHUTDOWN)
```

Comment:
This will help to avoid if on line 708 if (tm->ftm_state == FAB_TM_SHUTDOWN)  ?

if (tm->ftm_state == FAB_TM_SHUTDOWN)
{
    m0_mutex_unlock(&tm->ftm_endlock);
    return;
}

Reply:
Done